gamemodes/terrortown/entities/weapons/weapon_tttbase.lua

```diff
@@ -659,6 +668,99 @@ if CLIENT then
 		self.fCurrentSysTime = SysTime()
 	end
 
+	---
+	-- Adds a custom view model.
+	-- @note Multiple view models can be added, they are all rendered at once
+	-- @param string identifier The name of the added view model
+	-- @param ModelData modelData The model data table
+	-- @realm client
+	function SWEP:AddCustomViewModel(identifier, modelData)
+		self.VElements = self.VElements or {}
+		self.VElements[identifier] = modelData
+
+		modelbuilder.CreateModel(self, self.VElements[identifier])
+	end
+
+	---
+	-- Adds a custom world model.
+	-- @note Multiple world models can be added, they are all rendered at once
+	-- @param string identifier The name of the added world model
+	-- @param ModelData modelData The model data table
+	-- @realm client
+	function SWEP:AddCustomWorldModel(identifier, modelData)
+		self.WElements = self.WElements or {}
+		self.WElements[identifier] = modelData
+
+		modelbuilder.CreateModel(self, self.WElements[identifier])
+	end
+
+	---
+	-- Adds modifications to view model bones.
+	-- @param string identifier The identifier for this bone
+	-- @param BoneData boneData The bone data table
+	-- @realm client
+	function SWEP:ApplyViewModelBoneMods(identifier, boneData)
+		self.ViewModelBoneMods = self.ViewModelBoneMods or {}
+		self.ViewModelBoneMods[identifier] = boneData
+	end
+
+	---
+	-- Called straight after the view model has been drawn. This is called before
+	-- GM:PostDrawViewModel and WEAPON:PostDrawViewModel.
+	-- @param Entity viewModel Player's view model
+	-- @realm client
+	function SWEP:ViewModelDrawn(viewModel)
+		if self.ShowDefaultViewModel then
+			-- this resets the material to the default material
+			viewModel:SetMaterial("")
+		else
+			-- view model resets to render mode 0 every frame so we just apply
+			-- a debug material to prevent it from drawing
+			viewModel:SetMaterial("vgui/hsv")
+		end
+
+		if not self.VElements then return end
+
+		modelbuilder.UpdateBonePositions(self, viewModel)
+
+		-- we build a render order because sprites need to be drawn after models
+		self.VRenderOrder = modelbuilder.BuildRenderOrder(self.VElements, self.VRenderOrder)
+
+		modelbuilder.Render(self, self.VRenderOrder, self.VElements, viewModel)
+	end
+
+	---
+	-- Called when we are about to draw the world model.
+	-- @realm client
+	function SWEP:DrawWorldModel()
+		local client = LocalPlayer()
+
+		-- draw view model when spectating player
+		if client:GetObserverMode() == OBS_MODE_IN_EYE
+			and client:GetObserverTarget() == self:GetOwner()
+		then return end
+
+		if self.ShowDefaultWorldModel then
+			self:DrawModel()
+		end
+
+		if not self.WElements then return end
+
+		-- we build a render order because sprites need to be drawn after models
+		self.WRenderOrder = modelbuilder.BuildRenderOrder(self.WElements, self.WRenderOrder)
```

Comment:
I have to cash it somewhere to improve performance. And I thought it would be cleaner to have this outside the rendering code. But I don't care too much. Imho it is nicer that way, because generating a render order is not really a part of rendering it

Reply:
Well its much cleaner because it is something the renderer has to take into account and no one else has to know about this. Also you can still cache it somewhere on the entity.

Reply:
If you favor it, I can move it. I don't care that much

Reply:
Hm, as this function both renders the world and the viewmodel, I have to pass the correct render order as an argument. So I'm not sure how I should implement your suggestion

Reply:
Just move the function call inside the function. There are no dependencies on any other parameters. Or what are you talking about?

Reply:
How would I cache it though? Should I create the cache outside the render function and pass the table reference? Because I need one view model cache and one world model cache

Reply:
Well you can still just access the wep entity as before just store it in there or make a lookup table.

Reply:
yes, but both have to be called differently. How do I do this? pass two difference references? pass a string identifier?